linkerd/app/outbound/src/ingress.rs

```diff
@@ -79,142 +88,143 @@ impl<H> Outbound<H> {
-    type Key = Target;
+        // Route requests with destinations that can be discovered via the `l5d-dst-override` header
+        // through a logical stack and route requests without the `l5d-dst-override` header through
+        // the endpoint stack.
+        http_logical
+            .push_switch(
+                |(profile, Http { target, version }): (Option<profiles::Receiver>, _)| {
+                    // If the target did not include an override header, build an endpoint stack
+                    // with the original destination address (ignoring all headers, etc).
+                    if let Target::Forward(OrigDstAddr(addr)) = target {
+                        return Ok(svc::Either::B(http::Endpoint {
+                            addr: Remote(ServerAddr(addr)),
+                            metadata: Metadata::default(),
+                            logical_addr: None,
+                            protocol: version,
+                            opaque_protocol: false,
+                            tls: tls::ConditionalClientTls::None(
+                                tls::NoClientTls::IngressWithoutOverride,
+                            ),
+                        }));
+                    }
+
+                    // Otherwise, if a profile was discovered, use it to build a logical stack.
+                    if let Some(profile) = profile {
+                        let addr = profile.borrow().addr.clone();
+                        if let Some(logical_addr) = addr {
+                            return Ok(svc::Either::A(http::Logical {
+                                profile,
+                                logical_addr,
+                                protocol: version,
+                            }));
+                        }
+                    }
+
+                    // Otherwise, the override header was present but no profile information could
+                    // be discovered, so fail the request.
+                    Err(ProfileRequired)
+                },
+                http_endpoint,
+            )
+            .push(profiles::discover::layer(profiles, move |h: Http| {
+                // Lookup the profile if the override header was set and it is in the configured
+                // profile domains. Otherwise, profile discovery is skipped.
+                if let Target::Override(dst) = h.target {
+                    if profile_domains.matches(dst.name()) {
+                        return Ok(profiles::LookupAddr(dst.into()));
+                    }
+                }
 
-    fn recognize(&self, req: &http::Request<B>) -> Result<Self::Key, Error> {
-        let dst = http_request_l5d_override_dst_name_addr(req).map_err(|_| DstOverrideRequired)?;
-        Ok(Target {
-            dst,
-            version: self.0.protocol,
-        })
+                tracing::debug!(
+                    domains = %profile_domains,
+                    "Address not in a configured domain",
+                );
```

Comment:
it looks like this will be logged even if the target is a `Target::Forward`, which i don't think is correct --- should this be moved inside the `if let`? 

we could add a separate log line when it's a forward, like:
```rust
            if let Target::Override(dst) = h.target {
                if profile_domains.matches(dst.name()) {
                    return Ok(profiles::LookupAddr(dst.into()));
                }

                tracing::debug!(
                    domains = %profile_domains,
                    "Address not in a configured domain",
                );
            } else {
                tracing::trace!("No `{}` address; skipping discovery", DST_OVERRIDE_HEADER);
            }
```